casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -123,11 +157,48 @@ object Validate {
       val trueParents = ProtoUtil.chooseNonConflicting(estimate, genesis, dag).map(_.blockHash)
 
       if (bParents == trueParents)
-        true.pure[F]
+        Applicative[F].pure(Right(Valid))
       else
-        Log[F].warn(
-          ignore(b, "block parents did not match estimate based on justification.")
-        ) *> false.pure[F]
+        for {
+          _ <- Log[F].warn(
+                ignore(b, "block parents did not match estimate based on justification."))
+        } yield Left(InvalidParents)
+    }
+  }
+
+  /*
+   * When we switch between equivocation forks for a slashed validator, we will potentially get a
+   * justification regression that is valid. Note we can't immediately drop the slashed validator
+   * from the justifications as it might produce blocks on top of the equivocation that we
+   * need to check for justification regressions.
+   */
+  def justificationRegressions[F[_]: Applicative: Log](
+      b: BlockMessage,
+      genesis: BlockMessage,
+      dag: BlockDag): F[Either[BlockStatus, BlockStatus]] = {
+    val latestMessagesOfBlock = ProtoUtil.toLatestMessages(b.justifications)
+    val latestMessagesOfLatestMessagesForSender =
+      dag.latestMessagesOfLatestMessages.getOrElse(b.sender, latestMessagesOfBlock)
+    val containsJustificationRegression =
+      latestMessagesOfBlock.zip(latestMessagesOfLatestMessagesForSender).exists {
```

Comment:
Is `zip` guaranteed to associate pairs with the same key? I don't see any such guarantee in the docs and in fact it says "Note: might return different results for different runs, unless the underlying collection type is ordered." I think it's better to play it safe and write something like
```scala
latestMessagesOfBlock.exists {
  case (validator, currentBlockJustificationHash) =>
    //I'm actually not entirely sure what to do if `validator` is not in the other map, you can think it through though
    val previousBlockJustificationHash = latestMessagesOfLatestMessagesForSender.getOrElse(validator, ???)
    ...
}
```

Reply:
Great catch. It would mean that the previous latest message should have been already caught as invalid for missing a justification pointer or a new validator got rotated in (in which case just set it to the currentBlockJustificationHash as that is the "first block" from that validator).